triplestore-jena-memory/src/main/java/zone/cogni/asquare/triplestore/jenamemory/LocalTdbRdfStoreService.java

```diff
@@ -1,317 +1,418 @@
-    return executeInLock(Lock.READ, () -> {
-      try (QueryExecution queryExecution = bindings.asMap().isEmpty() ? QueryExecutionFactory.create(query, model)
-                                                                      : QueryExecutionFactory.create(query, model, bindings)
-      ) {
-        return queryExecution.execAsk();
-      }
-      catch (RuntimeException e) {
-        log.error("Query failed: {}", query);
-        throw e;
-      }
-    });
+  public <R> R executeSelectQuery(final String query, final JenaResultSetHandler<R> resultSetHandler, final String context) {
+    if (log.isDebugEnabled()) {
+      log.debug("Select {} \n{}", context == null ? "" : "--- " + context + " --- ", query);
+    }
+    return callQueryExecution(() -> createQueryExecution(query), resultSetHandler, context);
   }
 
   @Override
-  public Model executeConstructQuery(Query query, QuerySolutionMap bindings) {
-    return executeInLock(Lock.READ, () -> {
-      try (QueryExecution queryExecution = QueryExecutionFactory.create(query, model, bindings)) {
-        if (log.isTraceEnabled()) log.trace("Running construct query: \n{}", query);
-        return queryExecution.execConstruct();
-      }
-      catch (RuntimeException e) {
-        log.error("Query failed: {}", query);
-        throw e;
+  public boolean executeAskQuery(final Query query, final QuerySolutionMap bindings) {
+    validateReadiness();
+    return Txn.calculateRead(dataset, () -> {
+      try (final QueryExecution queryExecution = createQueryExecution(query, bindings)) {
+        return safeQuery(queryExecution::execAsk, queryExecution);
       }
     });
   }
 
   @Override
-  public Model executeConstructQuery(String query) {
-    return executeInLock(Lock.READ, () -> {
-      try (QueryExecution queryExecution = QueryExecutionFactory.create(query, model)) {
-        log.debug("Running construct query: \n{}", query);
-        return queryExecution.execConstruct();
-      }
-      catch (RuntimeException e) {
-        log.error("Query failed: {}", query);
-        throw e;
-      }
-    });
+  public Model executeConstructQuery(final Query query, final QuerySolutionMap bindings) {
+    if (log.isDebugEnabled()) {
+      log.debug("Running construct query: \n{}\nbindings: {}", query, bindings);
+    }
+    return callQueryConstruct(() -> createQueryExecution(query, bindings));
   }
 
   @Override
-  public void executeUpdateQuery(String updateQuery) {
-    executeInLock(Lock.WRITE, () -> UpdateAction.parseExecute(updateQuery, model));
+  public Model executeConstructQuery(final String query) {
+    if (log.isDebugEnabled()) {
+      log.debug("Running construct query: \n{}", query);
+    }
+    return callQueryConstruct(() -> createQueryExecution(query));
   }
 
-  @Override
-  public boolean graphExists(String graphUri) {
-    return datasetGraph.containsGraph(NodeFactory.createURI(graphUri));
+  public Model constructAllTriples() {
```

Comment:
Is this used? It seems to be a replacement for getModel. but the idea behind getModel might be flawed (except for in memory RdfStoreService... where it could be useful to get out the processed result.

Reply:
another option could be to have getModel do executeConstructQuery? or even copying the default model into a new model.

Reply:
ok, makes sense. thanks!